### ISSUE TITLE
Error in creating Pmat

### ISSUE THREAD

**Issue by Zepeng-Mu** [r3fang/SnapATAC#19]
Hi,

I am trying to generate a cell-by-peak matrix. Instead of running MCAS in snapATAC, I used the narrow peaks that I generated before with MACS2, and converted it to GRanges.

```
np <- read.table("5226cells_insertions_peaks.narrowPeak", header = F, stringsAsFactors = F)
npGr <- GRanges(np[, 1], IRanges(np[, 2], np[, 3]))
```

It looks OK:

```
GRanges object with 336764 ranges and 0 metadata columns:
           seqnames              ranges strand
              <Rle>           <IRanges>  <Rle>
       [1]     chr1         10045-10207      *
       [2]     chr1         16173-16419      *
       [3]     chr1       564481-566092      *
       [4]     chr1       566431-567376      *
       [5]     chr1       567520-568613      *
       ...      ...                 ...    ...
  [336760]     chr9 141014021-141014201      *
  [336761]     chr9 141014493-141014917      *
  [336762]     chr9 141015618-141015870      *
  [336763]     chr9 141029042-141029331      *
  [336764]     chr9 141074103-141074361      *
  -------
seqinfo: 22 sequences from an unspecified genome; no seqlengths
```

But when I run 
```
hcc <- createPmat(obj = hcc,
                  peak = npGr,
                  num.cores = 3)
```
I got the error

```
Error in .M.kind(x) : not yet implemented for matrix with typeof NULL
```

**Comment by Zepeng-Mu** [comment]
Am I supposed to use `addPmatToSnap` first and then call `createPmat`?

**Comment by r3fang** [comment]
Hi Zepeng, 

No, you don't need to run `addPmatToSnap` when running `createPmat`. This is an error that never happened before. Can you show me the following outcome:

```R
> hcc
> hcc$file[1:10]
```
thanks!

**Comment by Zepeng-Mu** [comment]
Sure.

```
> hcc
number of barcodes: 3377
number of bins: 0
number of genes: 0
number of peaks: 0
```

```
> hcc@file[1:10]
 [1] "/project2/yangili1/zpmu/scATAC-seq/data/snapatac/5226cell_named.snap"
 [2] "/project2/yangili1/zpmu/scATAC-seq/data/snapatac/5226cell_named.snap"
 [3] "/project2/yangili1/zpmu/scATAC-seq/data/snapatac/5226cell_named.snap"
 [4] "/project2/yangili1/zpmu/scATAC-seq/data/snapatac/5226cell_named.snap"
 [5] "/project2/yangili1/zpmu/scATAC-seq/data/snapatac/5226cell_named.snap"
 [6] "/project2/yangili1/zpmu/scATAC-seq/data/snapatac/5226cell_named.snap"
 [7] "/project2/yangili1/zpmu/scATAC-seq/data/snapatac/5226cell_named.snap"
 [8] "/project2/yangili1/zpmu/scATAC-seq/data/snapatac/5226cell_named.snap"
 [9] "/project2/yangili1/zpmu/scATAC-seq/data/snapatac/5226cell_named.snap"
[10] "/project2/yangili1/zpmu/scATAC-seq/data/snapatac/5226cell_named.snap"
```
This is the snap file I am using.
This error also happens when I try `creatGmat`.

**Comment by r3fang** [comment]
Sure! createGmat and createPmat is basically the same function with minor difference.
Let's figure this out step by step:

First, can you run this step and let me know the output?

```R
> reads.gr = extractReads(hcc@barcode[1:10], hcc@file[1:10], do.par = TRUE, num.cores = 1);
> reads.gr
```

If step 1 is successful, run this

```R
> library(plyr);
> library(GenomicRanges);
> ov = findOverlaps(reads.gr, npGr);
> idx = match(
			paste(reads.gr$file[queryHits(ov)], reads.gr$barcode[queryHits(ov)]),
			paste(hcc@file, hcc@barcode)
			);
> idy = subjectHits(ov);		
> ind = data.frame(idx,idy);
> ind  = plyr::count(ind, vars = c("idx", "idy"));
> head(ind);
```
Please let me know the outcome!

**Comment by Zepeng-Mu** [comment]
There is no error.

```
> head(ind)
  idx  idy freq
1   1    3    1
2   1    4    1
3   1   69    1
4   1 1732    1
5   1 2047    1
6   1 3339    1
```

**Comment by r3fang** [comment]
can you try this:

```R
library(parallel);
library(plyr);
library(Matrix);

obj = hcc;
ncell.chunk = 20;
peaks = npGr;
num.cores=5;

id = seq(nrow(obj));
id.ls = split(id, ceiling(seq(id)/ncell.chunk));
barcodes = obj@barcode;
files = obj@file;
ind.ls = mclapply(id.ls, function(x) {
    reads.gr = extractReads(barcodes[x], files[x], do.par = FALSE, num.cores = 1)
    ov = findOverlaps(reads.gr, peaks)
    idx = match(paste(reads.gr$file[queryHits(ov)], reads.gr$barcode[queryHits(ov)]), paste(obj@file, obj@barcode))
    idy = subjectHits(ov)
    ind = data.frame(idx, idy)
    plyr::count(ind, vars = c("idx", "idy"))
}, mc.cores = num.cores);

ind = Reduce(rbind, ind.ls)
head(ind);
M = sparseMatrix(i = ind[, 1], j = ind[, 2], x = ind[,3], dims = c(length(barcodes), length(peaks)))
```

This might take a bit long. but let me know which step failed.

**Comment by Zepeng-Mu** [comment]
So `id.ls` is OK I think, but `ind.ls` is all NULL.

**Comment by r3fang** [comment]
Sorry that I made a mistake, can you try it again. Really sorry. 

```R
library(parallel);
library(plyr);
library(Matrix);

obj = hcc;
ncell.chunk = 20;
peaks = npGr;
num.cores=5;
barcodes=obj@barcode;
files=obj@file;

id = seq(nrow(obj));
id.ls = split(id, ceiling(seq(id)/ncell.chunk));
barcodes = obj@barcode;
files = obj@file;
ind.ls = mclapply(id.ls, function(x) {
    reads.gr = extractReads(barcodes[x], files[x], do.par = FALSE, num.cores = 1)
    ov = findOverlaps(reads.gr, peaks)
    idx = match(paste(reads.gr$file[queryHits(ov)], reads.gr$barcode[queryHits(ov)]), paste(obj@file, obj@barcode))
    idy = subjectHits(ov)
    ind = data.frame(idx, idy)
    plyr::count(ind, vars = c("idx", "idy"))
}, mc.cores = num.cores);

ind = Reduce(rbind, ind.ls)
head(ind);
M = sparseMatrix(i = ind[, 1], j = ind[, 2], x = ind[,3], dims = c(length(barcodes), length(peaks)))
```

**Comment by r3fang** [comment]
let me know what is in `ind.ls`

**Comment by Zepeng-Mu** [comment]
Well, it's still NULL.

**Comment by r3fang** [comment]
Before we solve this problem, did the following raise the error: 

Error in .M.kind(x) : not yet implemented for matrix with typeof NULL

```R
> ind = Reduce(rbind, ind.ls)
```

**Comment by Zepeng-Mu** [comment]
So this step actually does not.

**Comment by Zepeng-Mu** [comment]
Error occurs here:
```
M = sparseMatrix(i = ind[, 1], j = ind[, 2], x = ind[,3], dims = c(length(barcodes), length(peaks)))
```

**Comment by r3fang** [comment]
Now i know where the problem is.

Can you run this:

```R
    x = id.ls[[1]]
    reads.gr = extractReads(barcodes[x], files[x], do.par = FALSE, num.cores = 1)
    ov = findOverlaps(reads.gr, peaks)
    idx = match(paste(reads.gr$file[queryHits(ov)], reads.gr$barcode[queryHits(ov)]), paste(obj@file, obj@barcode))
    idy = subjectHits(ov)
    ind = data.frame(idx, idy)
    a = plyr::count(ind, vars = c("idx", "idy"))
    head(a);
``` 

what is in a?

**Comment by Zepeng-Mu** [comment]
```
> head(a)
  idx  idy freq
1   1    3    1
2   1    4    1
3   1   69    1
4   1 1732    1
5   1 2047    1
6   1 3339    1
```

**Comment by r3fang** [comment]
how about this?

```R
ind.ls = lapply(id.ls, function(x) {
    reads.gr = extractReads(barcodes[x], files[x], do.par = FALSE, num.cores = 1)
    ov = findOverlaps(reads.gr, peaks)
    idx = match(paste(reads.gr$file[queryHits(ov)], reads.gr$barcode[queryHits(ov)]), paste(obj@file, obj@barcode))
    idy = subjectHits(ov)
    ind = data.frame(idx, idy)
    plyr::count(ind, vars = c("idx", "idy"))
});
```

**Comment by r3fang** [comment]
what is in `ind.ls`

**Comment by Zepeng-Mu** [comment]
It's now something like
```
$`5`
      idx    idy freq
1      81   1552    2
2      81   3156    1
3      81   5005    1
4      81   7402    1
5      81   9999    1
6      81  12360    1
7      81  12590    1
```
and if I `ind=Reduce(rbind, ind.ls)` we have
```
> head(ind)
  idx  idy freq
1   1    3    1
2   1    4    1
3   1   69    1
4   1 1732    1
5   1 2047    1
6   1 3339    1
```

**Comment by r3fang** [comment]
intersting, it seems that mclapply failed, it will modify this function to make it possible for non-parallel computing. Will let you know when it is ready

**Comment by r3fang** [comment]
hi, before i update the package. can you try to see if `createPmatTest` works?

```R
createPmatTest <- function (obj, peaks, ncell.chunk = 20, do.par = TRUE, num.cores = 1){
    if (missing(obj)) {
        stop("obj is missing")
    }
    else {
        if (!is(obj, "snap")) {
            stop("obj is not a snap file")
        }
    }
    if (missing(peaks)) {
        stop("peaks is missing")
    }
    else {
        if (!is(peaks, "GRanges")) {
            stop("peaks is not a GRanges object")
        }
    }
    if (!is(ncell.chunk, "numeric")) {
        stop("ncell.chunk is not numeric object")
    }
    if (!is(do.par, "logical")) {
        stop("do.par is not logical object")
    }
    if (!is(num.cores, "numeric")) {
        stop("num.cores is not logical object")
    }
    if (do.par) {
        if (num.cores > 1) {
            if (num.cores == 1) {
                num.cores = 1
            }
            else if (num.cores > detectCores()) {
                num.cores <- detectCores() - 1
                warning(paste0("num.cores set greater than number of available cores(",
                  parallel::detectCores(), "). Setting num.cores to ",
                  num.cores, "."))
            }
        }
        else if (num.cores != 1) {
            num.cores <- 1
        }
    }
    id = seq(nrow(obj))
    id.ls = split(id, ceiling(seq(id)/ncell.chunk))
    barcodes = obj@barcode
    files = obj@file
    if ((x = length(barcodes)) != (y = length(files))) {
        stop("obj has different length of barcode and file, recreate the snap object again!")
    }
    ind.ls = lapply(id.ls, function(x) {
        reads.gr = extractReads(barcodes[x], files[x], do.par = FALSE,
            num.cores = 1)
        ov = findOverlaps(reads.gr, peaks)
        idx = match(paste(reads.gr$file[queryHits(ov)], reads.gr$barcode[queryHits(ov)]),
            paste(obj@file, obj@barcode))
        idy = subjectHits(ov)
        ind = data.frame(idx, idy)
        plyr::count(ind, vars = c("idx", "idy"))
    })
    ind = Reduce(rbind, ind.ls)
    rm(ind.ls)
    gc()
    obj@pmat = sparseMatrix(i = ind[, 1], j = ind[, 2], x = ind[,
        3], dims = c(length(barcodes), length(peaks)))
    obj@peak = peaks
    rm(ind)
    gc()
    return(obj)
}

library(GenomicRanges)
library(SnapATAC)
library(parallel)
library(plyr);
a <- createPmat(obj = hcc, peak = npGr, num.cores = 1);
a
```

**Comment by Zepeng-Mu** [comment]
Hi, so I tried several times, it actually cases the R session to abort.

**Comment by r3fang** [comment]
Sorry i was wrong:

```
a <- createPmatTest(obj = hcc, peak = npGr, num.cores = 1);
```

If you can test wether `createPmatTest` works, it would be very helpful!

**Comment by Zepeng-Mu** [comment]
Hi, I actually used crearpamttest, I do not know if it has sokething to do
with the cluster I am using

Rongxin Fang <notifications@github.com> 于 2019年5月7日周二 19:24写道：

**Comment by r3fang** [comment]
are you able to solve this problem? Please let me know

**Comment by Zepeng-Mu** [comment]
Still not, let me try to run it on a personal computer.

**Comment by Zepeng-Mu** [comment]
Hi, I just realized, can you send me your R `sessionInfo` so that I could try to replicate it to see if I can run the method? Thanks, this is a great package that I hope I can try more of its features

**Comment by r3fang** [comment]
Sure, here is my R session info for machine 1

```
R version 3.4.0 (2017-04-21)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS release 6.6 (Final)

Matrix products: default
BLAS: /opt/R/lib64/R/lib/libRblas.so
LAPACK: /opt/R/lib64/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

loaded via a namespace (and not attached):
[1] compiler_3.4.0
```

and machine 2

```
> sessionInfo()
R version 3.5.2 (2018-12-20)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS: /mnt/silencer2/home/r3fang/software/R-3.5.2/lib/libRblas.so
LAPACK: /mnt/silencer2/home/r3fang/software/R-3.5.2/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8
 [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8
 [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

loaded via a namespace (and not attached):
[1] compiler_3.5.2
```

**Comment by Zepeng-Mu** [comment]
Well, I tried using 48G memory and 4 cores. I also filtered out cells with too many reads. Now I do not get the error anymore. I guess it's mainly a memory issue, there is a discussion on [stackoverflow](https://stackoverflow.com/questions/20674538/mclapply-returns-null-randomly) that also helped me a little bit.
I am really excited to try the package on cell-by-peak matrix!
Thanks (closing this for now)
